C++: Extend jump-to-def support to template instantiations.

criemen opened the pull request:
This PR extends developers ability to use jump-to-def in C/C++ files opened in the VSCode extension.
Before, jump-to-def we did not display definitions for code in template instantiations.

Furthermore, this PR fixes a bug, as the list of all references to a location did not include references inside template instantiations.

### Details ###
As part of the PR, this old comment is removed:
```
    // exclude results from template instantiations, as:
    // (1) these dependencies will often be caused by a choice of
    // template parameter, which is non-local to this part of code; and
    // (2) overlapping results pointing to different locations will
    // be very common.
    // It's possible we could allow a subset of these dependencies
    // in future, if we're careful to ensure the above don't apply.
```
Unfortunately, due to a limited history in the git repo, it is not clear when that comment was introduced.
* re (1): Non-locality is kinda the point of jump-to-def queries, so that should not be an issue.
* re (2): That is an issue on LGTM.com, VSCode has first-class support for a location having multiple definitions.

In the extension, the multiple definitions show up in a popup like this:
![2020-10-21-164446_751x334_scrot](https://user-images.githubusercontent.com/70492/96774469-4ff82380-13e6-11eb-99bb-4455774e9b0c.png)


### Questions from my side ###

* I need guidance on how to write test for these changes, as the interplay of the query and VSCode cannot be easily tested.
  So far, the one existing test has been adapted to test the old (LGTM.com) behaviour.

* Does this PR need some documentation/changelog entry?
  If yes, where?
  The change is user-visible, but only using the vscode extension and does not happen with a vscode extension release.
* I added a bit more context information to the top-level queries in ql/cpp/src, somebody needs to verify that it is correct/ the only places we use these queries indeed is VSCode or lgtm.com.

CC @github/codeql-core because the PR contains changes visible to VSCode extension users

jbj commented:
> Unfortunately, due to a limited history in the git repo, it is not clear when that comment was introduced.

Information on the pre-history of this repo is now in our team repository readme. See PR 172 there.

criemen commented:
Thank you! 
Checking the history of the comment does not provide more information, though.

geoffw0 commented:
I'm pretty sure I wrote that comment.  At the time I'd been tasked with cleaning up the LGTM jump-to-definition links, and LGTM handled overlapping links very poorly.

I think the concern motivating (1) was that, for example, the existence of a `std::vector<MyClass>` somewhere in your program would cause jump-to-definition on the template parameter `T`s inside the `std::vector` header file to take you to `MyClass`.

I understand (2) simply does not apply for VSCode, as it has an interface for displaying overlapping results.

criemen commented:
Thank you! I think we can live with (1), as the UI quite clearly lists all possibilities for T. It would be super confusing if we randomly picked one instance of T, though.

jbj commented on cpp/ql/src/definitions.qll:
```diff
@@ -197,14 +201,12 @@ Top definitionOf(Top e, string kind) {
-    // It's possible we could allow a subset of these dependencies
-    // in future, if we're careful to ensure the above don't apply.
-    not e.isFromTemplateInstantiation(_)
+    (
+      includeTemplateInstantiations = true
+      or
+      includeTemplateInstantiations = false and
+      not e.isFromTemplateInstantiation(_)
+    )
```
I think we should move the `isFromTemplateInstantiation` check up to the caller for the sake of performance. I'll go into more detail over Zoom later today.

criemen commented:
As discussed over Zoom, the predicate `definitionsOf` still contains plenty of duplicates for code in template instantiations.
For all elements in templates, we link from that element to its definition, even if all of these links share the same source and destination.
This happens quite a lot, for example for parameters of a function in a template, or for regular variables that are declared and used within a template.

In light of these duplicates, VSCode is doing the right thing on the client-side by ignoring multiple links, and not showing the popup when it is not needed.

We cannot easily divide out this equivalence relation, as it is not clear how to choose a representative for each equivalence class.
Changing the predicate to operate on `Location` instead of `Top` is not easy either, as that breaks the VSCode extension (and possibly LGTM.com, too).

To evaluate the impact of these duplicates on performance, we are running a CPP-differences job:
https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1515/

jbj commented on cpp/ql/src/localDefinitions.ql:
```diff
@@ -13,4 +13,4 @@ external string selectedSourceFile();
 from Top e, Top def, string kind
 where def = definitionOf(e, kind) and e.getFile() = getEncodedFile(selectedSourceFile())
-select e, def, kind
+select e.getLocation(), def.getLocation(), kind
```
I think you committed this line by accident

jbj commented on cpp/ql/src/definitions.qll:
```diff
@@ -196,15 +197,7 @@ Top definitionOf(Top e, string kind) {
     // exclude nested macro invocations, as they will overlap with
     // the top macro invocation.
-    not exists(e.(MacroAccess).getParentInvocation()) and
-    // exclude results from template instantiations, as:
-    // (1) these dependencies will often be caused by a choice of
-    // template parameter, which is non-local to this part of code; and
-    // (2) overlapping results pointing to different locations will
-    // be very common.
-    // It's possible we could allow a subset of these dependencies
-    // in future, if we're careful to ensure the above don't apply.
```
Please move this comment to `definitions.ql`, noting that `not e.isFromTemplateInstantiation(_)` is there due to a limitation in LGTM.

criemen commented:
I addressed the review.
The CPP-Differences jobs looks good in my eyes:
https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1517//artifact/analysis_time.html
